server/src/services/tooljet_db.service.ts

```diff
@@ -530,17 +756,175 @@ export class TooljetDbService {
+  ) {
+    const valid_referenced_table_details = await this.manager.find(InternalTable, {
+      where: {
+        organizationId: organisation_id,
+        ...(type === 'TABLENAME' && { tableName: In(referenced_table_list) }),
+        ...(type === 'TABLEID' && { id: In(referenced_table_list) }),
+      },
+      select: ['tableName', 'id'],
+    });
+
+    const referenced_tables_info = {};
+    const validReferencedTableSet = new Set(
+      valid_referenced_table_details.map((referenced_table_detail) => {
+        if (type === 'TABLEID') {
+          referenced_tables_info[referenced_table_detail.id] = referenced_table_detail.tableName;
+          return referenced_table_detail.id;
+        }
+        referenced_tables_info[referenced_table_detail.tableName] = referenced_table_detail.id;
+        return referenced_table_detail.tableName;
+      })
+    );
+
+    const invalid_tables = [];
+    const is_all_tables_exist = referenced_table_list.every((referenced_table) => {
+      if (validReferencedTableSet.has(referenced_table)) return true;
+      invalid_tables.push(referenced_table);
+      return false;
+    });
+
+    if (!is_all_tables_exist) {
+      const errorMessage =
+        type === 'TABLEID'
+          ? 'Some tables used in Foreign key was not found'
+          : `Tables: ${invalid_tables.join(',')} - used for Foreign key reference was not found`;
+      throw new BadRequestException(errorMessage);
+    }
+    return referenced_tables_info;
+  }
+
+  private async createForeignKey(organizationId: string, params) {
+    const { table_name, foreign_keys } = params;
+    if (!foreign_keys?.length) throw new BadRequestException('Foreign key details are missing');
+
+    const internalTable = await this.manager.findOne(InternalTable, {
+      where: { organizationId: organizationId, tableName: table_name },
+    });
+    if (!internalTable) throw new NotFoundException('Internal table not found: ' + table_name);
+
+    let referenced_tables_info = {};
+    const referenced_table_list = foreign_keys.map((foreign_key) => foreign_key.referenced_table_name);
+    referenced_tables_info = await this.fetchAndCheckIfValidForeignKeyTables(
+      referenced_table_list,
+      organizationId,
+      'TABLENAME'
+    );
+
+    const tjdbQueryRunner = this.tooljetDbManager.connection.createQueryRunner();
+    await tjdbQueryRunner.connect();
+    await tjdbQueryRunner.startTransaction();
+
+    try {
+      const foreignKeys = this.prepareForeignKeyDetailsJSON(foreign_keys, referenced_tables_info).map(
+        (foreignkeydetail) => new TableForeignKey({ ...foreignkeydetail })
+      );
+      await tjdbQueryRunner.createForeignKeys(internalTable.id, foreignKeys);
+
+      await tjdbQueryRunner.commitTransaction();
+      await this.tooljetDbManager.query("NOTIFY pgrst, 'reload schema'");
+      await tjdbQueryRunner.release();
+      return { statusCode: 200, message: 'Foreign key relation created successfully!' };
```

Comment:
Will advise to have request-response specific context only at the controller. It is better to return a boolean, check the result at the service (business logic) and give apt response code and message (request-response logic) at the controller.